src/index.js

```diff
@@ -31,38 +33,69 @@ class ESLintWebpackPlugin {
       extensions: arrify(this.options.extensions),
     };
 
-    if (options.lintDirtyModulesOnly) {
-      const dirtyFileWatcher = new DirtyFileWatcher(
-        options.files,
-        options.extensions
-      );
-
-      /* istanbul ignore next */
-      compiler.hooks.watchRun.tapPromise(
-        'ESLintWebpackPlugin',
-        async (runCompiler) => {
-          const files = dirtyFileWatcher.getDirtyFiles(
-            runCompiler.fileTimestamps
-          );
-
-          if (files.length > 0) {
-            await linter({ ...options, files }, runCompiler);
+    /**
+     * @param {Compiler} runCompiler
+     */
+    const run = async (runCompiler) => {
+      const wanted = parseFoldersToGlobs(options.files, options.extensions);
+      /** @type {string[]} */
+      const files = [];
+      /** @type {import('./linter').Linter} */
+      let lint;
+      /** @type {import('./linter').Reporter} */
+      let report;
+
+      try {
+        ({ lint, report } = linter(options));
+      } catch (e) {
+        runCompiler.hooks.thisCompilation.tap(ESLINT_PLUGIN, (compilation) =>
+          compilation.errors.push(e)
+        );
+        return;
+      }
+
+      runCompiler.hooks.thisCompilation.tap(ESLINT_PLUGIN, (compilation) => {
+        compilation.hooks.succeedModule.tap(ESLINT_PLUGIN, (module) => {
+          // @ts-ignore
+          const file = module.resource;
+
+          if (file && micromatch.isMatch(file, wanted)) {
+            // Queue file for linting.
+            files.push(file);
           }
-        }
-      );
-    } else {
-      compiler.hooks.run.tapPromise('ESLintWebpackPlugin', (runCompiler) => {
-        return linter(options, runCompiler);
+        });
+
+        runCompiler.hooks.make.tapPromise(ESLINT_PLUGIN, async () =>
+          lint(files)
+        );
+
+        runCompiler.hooks.emit.tapPromise(ESLINT_PLUGIN, async () => {
```

Comment:
Why `emit`? I wrote above that we should use

Reply:
I'm learning as I go. I didn't know where your snippet went or how to use it. Would it go here?

Reply:
`compilation.hooks.additionalChunkAssets` fires really early and its sync. I need an async event before it and after all modules are found. `compilation.hooks.afterSeal` is async, and allows me to get the async stuff completed... but it fires after `additionalChunkAssets`. :|

Reply:
hm, we should not write assets in `emit` hook, we can use `compier.outputFilesystem` in this case

Reply:
done.